File: JDBCProj/src/util/jdbcUtil.java
package util;

import com.alibaba.druid.pool.DruidDataSourceFactory;

import javax.sql.DataSource;
import java.io.IOException;
import java.io.InputStream;
import java.sql.*;
import java.util.Properties;

public class jdbcUtil {

    private static DataSource dataSource;

    public static Connection getConnection() throws SQLException {
        if (dataSource==null){
            try {
                Properties properties=new Properties();
                InputStream in=jdbcUtil.class.getClassLoader().getResourceAsStream("druid.properties");
                properties.load(in);
                in.close();
                dataSource= DruidDataSourceFactory.createDataSource(properties);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
        Connection connection = dataSource.getConnection();
        return connection;
    }

    /**
     *获得数据库连接
     * @return 连接对象
     * @throws IOException
     * @throws ClassNotFoundException
     * @throws SQLException
     */
    public static Connection getConnectionOld() throws IOException, ClassNotFoundException, SQLException {
        Properties properties=new Properties();
        /**
         * 通过properties配置文件存储驱动和数据库信息，并通过IO流读取到properties集合中
         */
        InputStream inputStream=jdbcUtil.class.getClassLoader().getResourceAsStream("jdbc.properties");
        properties.load(inputStream);
        inputStream.close();

        /**
         * 从集合中获取驱动和数据库信息供类加载和驱动获得连接
         */
        String driverClassName=properties.getProperty("driverClassName");
        String url=properties.getProperty("url");
        String user=properties.getProperty("user");
        String password=properties.getProperty("password");

        /**
         * 自动注册驱动，并通过驱动管理器获得数据库连接
         */
        Class.forName(driverClassName);
        Connection connection= DriverManager.getConnection(url,user,password);
        return  connection;
    }

    /**
     *释放资源
     * @param connection 数据库连接
     */
    public static void close(Connection connection){
        close(connection,null);
    }

    /**
     * 释放资源
     * @param connection 数据库连接
     * @param statement 处理流
     */
    public static void close(Connection connection, Statement statement){
        close(connection,statement,null);
    }

    public static void close(Connection connection, Statement statement, ResultSet resultset){
        if (resultset!=null){
            try {
                resultset.close();
            } catch (SQLException e) {
                e.printStackTrace();
            }
        }

        if (statement!=null){
            try {
                statement.close();
            } catch (SQLException e) {
                e.printStackTrace();
            }
        }

        if (connection!=null){
            try {
                connection.close();
            } catch (SQLException e) {
                e.printStackTrace();
            }
        }
    }
}

File: SparkStreaming_AD/src/main/scala/util/JDBCUtil.scala
package util

import java.sql.{Connection, PreparedStatement, ResultSet}
import java.util.Properties

import com.alibaba.druid.pool.DruidDataSourceFactory
import javax.sql.DataSource

/**
 * Copyright(c) 2020-2021 sparrow All Rights Reserved
 * Project: SparkStreaming_AD
 * Package: util
 * ClassName: JDBCUtil
 *
 * @author 18729 created on date: 2020/11/2 11:20
 * @version 1.0
 * @since JDK 1.8
 */
object JDBCUtil {
  //初始化连接池
  var dataSource: DataSource = init()

  //初始化连接池方法
  def init(): DataSource = {
    val properties = new Properties()
    val config: Properties = PropertiesUtil.load("config.properties")
    properties.setProperty("driverClassName", "com.mysql.jdbc.Driver")
    properties.setProperty("url", config.getProperty("jdbc.url"))
    properties.setProperty("username", config.getProperty("jdbc.user"))
    properties.setProperty("password", config.getProperty("jdbc.password"))
    properties.setProperty("maxActive", config.getProperty("jdbc.datasource.size"))
    DruidDataSourceFactory.createDataSource(properties)
  }

  //获取MySQL连接
  def getConnection: Connection = {
    dataSource.getConnection
  }

  //执行SQL语句,单条数据插入
  def executeUpdate(connection: Connection, sql: String, params: Array[Any]): Int = {
    var rtn = 0
    var pstmt: PreparedStatement = null
    try {
      connection.setAutoCommit(false)
      pstmt = connection.prepareStatement(sql)
      if (params != null && params.length > 0) {
        for (i <- params.indices) {
          pstmt.setObject(i + 1, params(i))
        }
      }
      rtn = pstmt.executeUpdate()
      connection.commit()
      pstmt.close()
    } catch {
      case e: Exception => e.printStackTrace()
    }
    rtn
  }

  //判断一条数据是否存在
  def isExist(connection: Connection, sql: String, params: Array[Any]): Boolean = {
    var flag: Boolean = false
    var pstmt: PreparedStatement = null
    try {
      pstmt = connection.prepareStatement(sql)
      for (i <- params.indices) {
        pstmt.setObject(i + 1, params(i))
      }
      flag = pstmt.executeQuery().next()
      pstmt.close()
    } catch {
      case e: Exception => e.printStackTrace()
    }
    flag
  }

  //获取MySQL的一条数据
  def getDataFromMysql(connection: Connection, sql: String, params: Array[Any]): Long = {
    var result: Long = 0L
    var pstmt: PreparedStatement = null
    try {
      pstmt = connection.prepareStatement(sql)
      for (i <- params.indices) {
        pstmt.setObject(i + 1, params(i))
      }
      val resultSet: ResultSet = pstmt.executeQuery()
      while (resultSet.next()) {
        result = resultSet.getLong(1)
      }
      resultSet.close()
      pstmt.close()
    } catch {
      case e: Exception => e.printStackTrace()
    }
    result
  }

  //主方法,用于测试上述方法
  def main(args: Array[String]): Unit = {
    //1 获取连接
    val connection: Connection = getConnection
    //2 预编译SQL
    val statement: PreparedStatement = connection.prepareStatement("select * from user_ad_count where userid = ?")
    //3 传输参数
    statement.setObject(1, 'a')
    //4 执行sql
    val resultSet: ResultSet = statement.executeQuery()
    //5 获取数据
    while (resultSet.next()) {
      println("111:" + resultSet.getString(1))
    }
    //6 关闭资源
    resultSet.close()
    statement.close()
    connection.close()
  }
}
